lib/spack/spack/fetch_strategy.py

```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+        # type: (str) -> Optional[str]
+        try:
+            with working_dir(path):
+                check_commit_pattern = self.ref.refspec() + '^{commit}'
+                return self.git('rev-parse', '--verify', check_commit_pattern)
+        except ProcessError as e:
+            tty.debug('failed to find ref {0} locally: {1}'.format(self.ref, e))
+        return None
 
-        return '{0}{1}'.format(self.url, args)
+    def _do_fetch(self):
+        # type: () -> str
 
-    @_needs_stage
-    def fetch(self):
-        if self.stage.expanded:
-            tty.debug('Already fetched {0}'.format(self.stage.source_path))
-            return
+        verbosity_arg = '--verbose' if spack.config.get('config:debug') else '--quiet'
 
-        tty.debug('Cloning git repository: {0}'.format(self._repo_info()))
-
-        git = self.git
-        if self.commit:
-            # Need to do a regular clone and check out everything if
-            # they asked for a particular commit.
-            debug = spack.config.get('config:debug')
-
-            clone_args = ['clone', self.url]
-            if not debug:
-                clone_args.insert(1, '--quiet')
-            with temp_cwd():
-                git(*clone_args)
-                repo_name = get_single_file('.')
-                self.stage.srcdir = repo_name
-                shutil.move(repo_name, self.stage.source_path)
-
-            with working_dir(self.stage.source_path):
-                checkout_args = ['checkout', self.commit]
-                if not debug:
-                    checkout_args.insert(1, '--quiet')
-                git(*checkout_args)
+        submodules_arg = '--recurse-submodules={0}'.format(
+            'on-demand' if self.submodules else 'no')
 
-        else:
-            # Can be more efficient if not checking out a specific commit.
-            args = ['clone']
-            if not spack.config.get('config:debug'):
-                args.append('--quiet')
-
-            # If we want a particular branch ask for it.
-            if self.branch:
-                args.extend(['--branch', self.branch])
-            elif self.tag and self.git_version >= ver('1.8.5.2'):
-                args.extend(['--branch', self.tag])
-
-            # Try to be efficient if we're using a new enough git.
-            # This checks out only one branch's history
-            if self.git_version >= ver('1.7.10'):
-                if self.get_full_repo:
-                    args.append('--no-single-branch')
+        # We do not add --force, since we do *not* want to ever modify any tags.
+        with working_dir(self._cache_path()):
+            if self.get_full_repo:
+                trailing_args = ['--tags', self.url, '+refs/heads/*:refs/heads/*']
+            else:
+                trailing_args = [self.url, self.ref.updating_refspec()]
+            return self.git(
+                'fetch', verbosity_arg, submodules_arg,
+                *tuple(trailing_args))
```

Comment:
Why convert to a tuple here?

Reply:
If it works, great! I'll check that. I was under the impression python 2 (maybe 2.6?) strictly requires tuples to be used for positional splats.